### PR TITLE
Add expressions for creating triangular/square/curved waves

### PR DESCRIPTION
**Pull request by nyalldawson**
The new expression functions are:
- triangular_wave: Constructs triangular waves along the boundary of a geometry.
- square_wave: Constructs square/rectangular waves along the boundary of a geometry
- wave: Constructs rounded (sine-like) waves along the boundary of a geometry
- triangular_wave_randomized: Constructs randomized triangular waves along the boundary of a geometry
- square_wave_randomized: Constructs randomized square/rectangular waves along the boundary of a geometry.
- wave_randomized: Constructs randomized curved (sine-like) waves along the boundary of a geometry

Sponsored by North Road, thanks to SLYR

These are designed to unlock some fancy symbology effects when used with geometry generators! eg...

image

image

image

image

### CODE REVIEW

**Comment by nyalldawson** [comment]
> Have you considered using overloaded methods/expressions instead of the _randomized variants?

I flip-flopped back and forward here, but ultimately I think it'd be a mistake to try and jam them into a single function. Eg. I'd like to add a "tightness" parameter sometime to the curved wave function, which would then mean it'd have four arguments after the geometry, which would then clash with the minimum number of arguments required by the wave_randomized function. Safer just to keep them as separate functions.

File: src/core/geometry/qgsinternalgeometryengine.cpp [hunk 1]
```diff
@@ -1645,3 +1645,985 @@ QgsGeometry QgsInternalGeometryEngine::orientedMinimumBoundingBox( double &area,
...
+    geom = segmentizedCopy.get();
+  }
+
+  if ( QgsWkbTypes::geometryType( geom->wkbType() ) == QgsWkbTypes::LineGeometry )
+  {
+    return roundWavesAlongLine( static_cast< const QgsLineString * >( geom ), wavelength, amplitude, strictWavelength );
+  }
+  else
+  {
+    // polygon
```
Comment (nirvn) [inline src/core/geometry/qgsinternalgeometryengine.cpp:2502]: Where are we guarding ourselves from the abstract geometry pointer being a point geometry? Should we check for point type and return an empty abstract geometry here (and in the two other triangle/square private functions?

File: tests/src/core/testqgsexpression.cpp [hunk 2]
```diff
@@ -1329,6 +1329,24 @@ class TestQgsExpression: public QObject
       QTest::newRow( "affine_transform line XY" ) << "geom_to_wkt(affine_transform(geom_from_wkt('LINESTRING(1 0, 2 0)'), 0, 0, 90, 2, 1))" << false << QVariant( "LineString (0 2, 0 4)" );
       QTest::newRow( "affine_transform polygon XYZ" ) << "geom_to_wkt(affine_transform(geom_from_wkt('POLYGON((0 0, 0 1, 1 1, 1 0, 0 0))'), 0, 0, -90, 0.5, 0.5))" << false << QVariant( "Polygon ((0 0, 0.5 0, 0.5 -0.5, 0 -0.5, 0 0))" );
       QTest::newRow( "affine_transform point XY with translation on ZM" ) << "geom_to_wkt(affine_transform(geom_from_wkt('POINT(1 1)'), 0, 0, 0, 1, 1, 3, 4))" << false << QVariant( "PointZM (1 1 3 4)" );
+      QTest::newRow( "triangular_wave not geom" ) << "triangular_wave('g', 1, 2)" << true << QVariant();
```
Comment (nirvn) [inline tests/src/core/testqgsexpression.cpp:1332]: Maybe add an incompatible geometry test in there too (i.e. point)

**Review by nirvn — COMMENTED** [review]
Code looks good, feature is _super_ cool. Only one question raised.

File: src/core/geometry/qgsinternalgeometryengine.cpp (hunk 1 above)
Comment (nyalldawson) [reply src/core/geometry/qgsinternalgeometryengine.cpp:2502]: It's done in the public functions -- e.g. at https://github.com/qgis/QGIS/pull/45583/files/d9e3659879c8ab168302755a68bb13f561b2b68f#diff-17cd6748c845e0043f19426bb01e791a11d7a0310044693d2fcb4a08d5c5c4c9R2560
So here we are guaranteed to have line or polygon types (the public functions also handle the multipart types, so again we are guaranteed here to have a single-part, linear type). There's unit tests covering all this too for safety.

**Review by nirvn — APPROVED** [review]
Doh, that was staring me in the face...

File: tests/src/core/testqgsexpression.cpp (hunk 2 above)
Comment (nyalldawson) [reply tests/src/core/testqgsexpression.cpp:1332]: Ok, done

**Pull request merged**